akka-stream/src/main/scala/akka/stream/impl/fusing/FlatMapPrefix.scala

```diff
@@ -0,0 +1,156 @@
+          subSource.get.push(grab(in))
+        } else {
+          accumulated ::= grab(in)
+          if (accumulated.size == n) {
+            materializeFlow()
+          } else {
+            //gi'me some more!
+            pull(in)
+          }
+        }
+      }
+
+      override def onUpstreamFinish(): Unit = {
+        if (subSource.isDefined) {
+          subSource.get.complete()
+        } else {
+          materializeFlow()
+        }
+      }
+
+      override def onUpstreamFailure(ex: Throwable): Unit = {
+        if (subSource.isDefined) {
+          subSource.get.fail(ex)
+        } else {
+          //flow won't be materialized, so we have to complete the future with a failure indicating this
+          matPromise.failure(new NeverMaterializedException(ex))
+          super.onUpstreamFailure(ex)
+        }
+      }
+
+      override def onPull(): Unit = {
+        if (subSink.isDefined) {
+          //delegate to subSink
+          subSink.get.pull()
+        } else if (accumulated.size < n) {
+          pull(in)
+        } else if (accumulated.size == n) { //corner case for n = 0, can be handled in FlowOps
+          materializeFlow()
+        }
+      }
+
+      override def onDownstreamFinish(cause: Throwable): Unit = {
+        if (subSink.isEmpty) {
+          materializeFlow()
+        }
+        subSink.get.cancel(cause)
+      }
+
+      def materializeFlow(): Unit = {
+        val prefix = accumulated.reverse
+        accumulated = Nil
+        subSource = OptionVal.Some(new SubSourceOutlet[In]("subSource"))
+        subSource.get.setHandler {
+          new OutHandler {
+            override def onPull(): Unit = {
+              if (!isClosed(in) && !hasBeenPulled(in)) {
+                pull(in)
+              }
+            }
+
+            override def onDownstreamFinish(cause: Throwable): Unit = {
+              if (!isClosed(in)) {
+                cancel(in, cause)
+              }
+              //super.onDownstreamFinish(cause)
+            }
+          }
+        }
+        subSink = OptionVal.Some(new SubSinkInlet[Out]("subSink"))
+        subSink.get.setHandler {
+          new InHandler {
+            override def onPush(): Unit = {
+              push(out, subSink.get.grab())
+            }
+
+            override def onUpstreamFinish(): Unit = {
+              complete(out)
+            }
+
+            override def onUpstreamFailure(ex: Throwable): Unit = {
+              fail(out, ex)
+            }
+          }
+        }
+        val matVal = try {
+          val flow = f(prefix)
+          val runnableGraph = Source.fromGraph(subSource.get.source).viaMat(flow)(Keep.right).to(subSink.get.sink)
+          interpreter.subFusingMaterializer.materialize(runnableGraph)
+        } catch {
+          case NonFatal(ex) =>
+            matPromise.failure(new NeverMaterializedException(ex))
+            subSource = OptionVal.None
+            subSink = OptionVal.None
+            throw ex
```

Comment:
Use failStage instead to prevent extra log messages (plus points for adding an `EventFilter` to the test to check that no extra log message was thrown).

Reply:
can u please point me to an example of using such an event filter in tests?

Reply:
re. failStage, I need this exception to terminate this method's execution. this is easily achieved by throwing an exception, reverting to failStag will require introduction of a boolean var - which I'd be happy to avoid if possible

Reply:
Then there needs to be a try/catch around materializeFlow() which ultimately calls `failStage`.